File: algorithm/cpp/n-queens-ii.cpp
// 52. N-Queens II
// Difficulty: Hard

// Follow up for N-Queens problem.

// Now, instead outputting board configurations, return the total number of distinct solutions.

// Time : O(n!)
// Space: O(n)

class Solutions {
public:
  int totalNQueens(int n) {
    this->columns = vector<bool>(n, false);
    this->diags = vector<bool>(2 * n - 1, false);
    this->diags2 = vector<bool>(2 * n - 1, false);
    int count;
    vector<int> colQ(n, -1);
    totalNQueens(colQ, count, 0);
    return count;
  }
private:
  vector<bool> columns;
  vector<bool> diags;
  vector<bool> diags2;
  void totalNQueens(vector<int>& colQ, int& count, int row) {
    const int N = colQ.size();
    if (row == N) {
      ++count;
      return;
    }
    for (int j = 0; j < N; ++j) {
      if (!columns[j] && !diags[row - j + N - 1] && !diags2[row + j]) {
        colQ[row] = j;
        columns[j] = diags[row - j + N - 1] = diags2[row + j] = true;
        totalNQueens(colQ, count, row + 1);
        colQ[row] = -1;
        columns[j] = diags[row - j + N - 1] = diags2[row + j] = false;
      }
    }
  }
};

File: zZ/java/n-queens-ii.java
/**
 * N-Queens II
 * 
 * Follow up for N-Queens problem.
 * 
 * Now, instead outputting board configurations, return the total number of distinct solutions.
 */

public class Solution {
	public int totalNQueens(int n) {
		int[] result = new int[1];
		int[] columnPositionInRow = new int[n];
		Arrays.fill(columnPositionInRow, -1);
		totalNQueensHelper(row, n, columnPositionInRow, result);
		return result[0];
	}

	private totalNQueensHelper(int row, int n, int[] columnPositionInRow, int[] result) {
		if (row == n) {
			++result[0];
		} else {
			for (int col = 0; col < n; ++col) {
				if (isValid(columnPositionInRow, row, col)) {
					columnPositionInRow[row] = col;
					totalNQueensHelper(row + 1, n, columnPositionInRow, result);
					columnPositionInRow[row] = -1;
				}
			}
		}
	}

	private bool isValid(int[] columnPositionInRow, int row, int col) {
		for (int i = 0; i < columnPositionInRow.length; ++i) {
			if (columnPositionInRow[i] != -1) {
				if (columnPositionInRow[i] == col || row - i == col - columnPositionInRow[i]) {
					return false;
				}
			}
		}
		return true;
	}
}
